In [1]:
import pandas as pd

#### 1. Loading the  5 datasets from the "trips_db" Database

In [2]:

dimCity = pd.read_csv("dim_city.csv")
dimDate = pd.read_csv("dim_date.csv")
passengerSummary = pd.read_csv('C:/2 Code/11 Nov 2024/Codebasics CH 13 Cab/datasets/csv_files/fact_passenger_summary.csv')
repeatTrip = pd.read_csv("dim_repeat_trip_distribution.csv")
factTrips = pd.read_csv("fact_trips.csv")


In [3]:
passengerSummary.shape

(60, 5)

In [4]:
passengerSummary.head()

month city_id  new_passengers  repeat_passengers  total_passengers
0  2024-01-01    AP01            2513                650              3163
1  2024-01-01    CH01            3920                720              4640
2  2024-01-01    GJ01            2432               1184              3616
3  2024-01-01    GJ02            2089                544              2633
4  2024-01-01    KA01            1957                172              2129

In [5]:
print(dimCity.shape)
print(dimCity.head(11))

(10, 2)
  city_id      city_name
0    RJ01         Jaipur
1    UP01        Lucknow
2    GJ01          Surat
3    KL01          Kochi
4    MP01         Indore
5    CH01     Chandigarh
6    GJ02       Vadodara
7    AP01  Visakhapatnam
8    TN01     Coimbatore
9    KA01         Mysore


In [6]:
print(dimDate.shape)
print(dimDate.tail())

(182, 4)
           date start_of_month month_name day_type
177  2024-06-26     2024-06-01       June  Weekday
178  2024-06-27     2024-06-01       June  Weekday
179  2024-06-28     2024-06-01       June  Weekday
180  2024-06-29     2024-06-01       June  Weekend
181  2024-06-30     2024-06-01       June  Weekend


In [7]:
print(repeatTrip.shape)
print(repeatTrip.head(15))

(540, 4)
         month city_id trip_count  repeat_passenger_count
0   2024-01-01    AP01   10-Trips                       7
1   2024-01-01    AP01    2-Trips                     352
2   2024-01-01    AP01    3-Trips                     158
3   2024-01-01    AP01    4-Trips                      53
4   2024-01-01    AP01    5-Trips                      38
5   2024-01-01    AP01    6-Trips                      14
6   2024-01-01    AP01    7-Trips                      10
7   2024-01-01    AP01    8-Trips                      11
8   2024-01-01    AP01    9-Trips                       7
9   2024-01-01    CH01   10-Trips                      14
10  2024-01-01    CH01    2-Trips                     196
11  2024-01-01    CH01    3-Trips                     182
12  2024-01-01    CH01    4-Trips                      87
13  2024-01-01    CH01    5-Trips                     119
14  2024-01-01    CH01    6-Trips                      44


In [8]:
print(factTrips.shape)
factTrips.tail()

(425903, 8)


trip_id        date city_id passenger_type  \
425898  TRPLUC2403180b02b4d0  2024-03-18    UP01       repeated   
425899  TRPVAD2401032679e669  2024-01-03    GJ02       repeated   
425900  TRPJAI24022578e10280  2024-02-25    RJ01       repeated   
425901  TRPJAI240401e297ad20  2024-04-01    RJ01       repeated   
425902  TRPJAI24021056f0e951  2024-02-10    RJ01            new   

        distance_travelled(km)  fare_amount  passenger_rating  driver_rating  
425898                      12          134                 5              5  
425899                      12          114                 7              5  
425900                      26          479                 7             10  
425901                      27          361                 7             10  
425902                      37          666                 9             10

In [9]:
# Ensure the date column is in datetime format
factTrips['date'] = pd.to_datetime(factTrips['date'])

# Sort the DataFrame by the date column
factTrips_sorted = factTrips.sort_values(by='date')

# Display the first few rows of the sorted DataFrame

factTrips_sorted.head()


trip_id       date city_id passenger_type  \
212951  TRPJAI24010195e6afba 2024-01-01    RJ01       repeated   
153552  TRPCOI2401018e343b87 2024-01-01    TN01            new   
154215  TRPSUR2401012dded9ee 2024-01-01    GJ01       repeated   
374624  TRPJAI240101759605cc 2024-01-01    RJ01       repeated   
154625  TRPVIS2401017c9a8906 2024-01-01    AP01       repeated   

        distance_travelled(km)  fare_amount  passenger_rating  driver_rating  
212951                      15          217                 7              8  
153552                      20          183                 8              8  
154215                      10           99                 7              8  
374624                      25          337                 9              9  
154625                      13          144                 8             10

- Observations and insights so far:-
---
- "dimCity" dataset provides names of the cities in which the company is operating.
- There are 10 cities in which the company is operating.
---
- "dimDate" provdes date related info.
- I can't find any direct usage of the info in this dataset. Maybe I will find it elevent later on during analysis.
---
- "PasengerSumary" dataset contains the number of "new_passengers", "repeat_pasengers" and "total_passengers" for each month for each of the 10 cities starting from January 2024 to June 2024.
---
- "repeatTrips" contains info on the passengers who have taken more than 1 trip.
- Specifically, it contains data about how many passengers have taken 2,3,4....,10 trips. 
- Data above 10 trips does not exist.
- There are 9 entries for each of the 10 cities, for each of the 6 months. So total rowws = 9*10*6 = 540 rows.
---
- "factTrips" is the biggest dataset of the 5. It contains details of each individual trip over the 6 month period.
- It looks like the data isn't in chronological order. I will have to rearrange/ sort it.

#### 2. Loading the datasets in the "targets_db" Database

In [10]:
targetRating = pd.read_csv("city_target_passenger_rating.csv")
targetNewPassengers= pd.read_csv("monthly_target_new_passengers.csv")
targetTrips = pd.read_csv("monthly_target_trips.csv")

In [11]:
print(targetRating.shape)
print(targetRating)

(10, 2)
  city_id  target_avg_passenger_rating
0    CH01                         8.00
1    UP01                         7.25
2    AP01                         8.50
3    MP01                         8.00
4    RJ01                         8.25
5    KA01                         8.50
6    GJ01                         7.00
7    TN01                         8.25
8    KL01                         8.50
9    GJ02                         7.50


In [12]:
print(targetNewPassengers.shape)
print(targetNewPassengers.head(60))

(60, 3)
         month city_id  target_new_passengers
0   2024-05-01    GJ01                   1500
1   2024-05-01    GJ02                   1500
2   2024-03-01    GJ01                   2000
3   2024-05-01    UP01                   2000
4   2024-05-01    MP01                   2000
5   2024-01-01    AP01                   2500
6   2024-03-01    MP01                   2700
7   2024-06-01    CH01                   3000
8   2024-04-01    AP01                   2000
9   2024-01-01    KA01                   2000
10  2024-03-01    UP01                   3200
11  2024-02-01    AP01                   2500
12  2024-04-01    GJ02                   1500
13  2024-02-01    TN01                   1500
14  2024-01-01    MP01                   2700
15  2024-03-01    TN01                   1500
16  2024-05-01    KA01                   2000
17  2024-05-01    TN01                   1000
18  2024-02-01    GJ01                   2000
19  2024-04-01    CH01                   3000
20  2024-01-01    GJ02    

In [13]:
print(targetTrips.shape)
print(targetTrips.head())

(60, 3)
        month city_id  total_target_trips
0  2024-03-01    MP01                7000
1  2024-05-01    KA01                2500
2  2024-04-01    UP01               11000
3  2024-02-01    GJ02                6000
4  2024-05-01    KL01                9000


- Observations so far:-
---
- "targetRating" contains the target average passenger rating that Goodcabs aims to achieve for each city.
- This rating is used to monitor customer satisfaction.
---
-"targetPassengers" contains data about the target number of passengers Goodcabs aims to acquire in each of the 10 cities for each of the 6 months. 
- Hence total no. of rows = 10* 6 = 60.
---
- "targetTrips" contains data about total trips Goodcabs aims to achieve for each of the 10 cities in each of the 6 months.



#### Calculating metrics

1. Top 3 and bottom 3 cities by total trips over the 6 month period

In [14]:
import pandas as pd

# Aggregating total trips by city
city_trip_summary = factTrips.groupby('city_id')['trip_id'].count().reset_index()
city_trip_summary.rename(columns={'trip_id': 'total_trips'}, inplace=True)

# Sorting cities by total trips
city_trip_summary_sorted = city_trip_summary.sort_values(by='total_trips', ascending=False)

# Top 3 and Bottom 3 performing cities
top_3_cities = city_trip_summary_sorted.head(3)
bottom_3_cities = city_trip_summary_sorted.tail(3)

# Merging with city names for better readability
top_3_cities = top_3_cities.merge(dimCity, on='city_id', how='left')
bottom_3_cities = bottom_3_cities.merge(dimCity, on='city_id', how='left')

print("Top 3 Performing Cities:")
print(top_3_cities[['city_name', 'total_trips']])

print("\nBottom 3 Performing Cities:")
print(bottom_3_cities[['city_name', 'total_trips']])


Top 3 Performing Cities:
  city_name  total_trips
0    Jaipur        76888
1   Lucknow        64299
2     Surat        54843

Bottom 3 Performing Cities:
       city_name  total_trips
0  Visakhapatnam        28366
1     Coimbatore        21104
2         Mysore        16238


- Observations
- Jaipur, Lucknow and Surat had the highest number of trips 
- Visakhapatnam, Coimbatore and Mysore had the least number of trips

2. Average Fare Per Trip, Average Distance(in km) per trip and Fare per kilometer

In [15]:
#Calculate average fare per trip by city
city_fare_summary = factTrips.groupby('city_id').agg(
    total_fare= ('fare_amount','sum'),
    total_trips= ('trip_id','count'),
    total_distance = ('distance_travelled(km)','sum'),
).reset_index()

city_fare_summary['avg_fare_per_trip'] = city_fare_summary['total_fare']/ city_fare_summary['total_trips']
city_fare_summary['avg_fare_per_trip'] 
city_fare_summary['avg_dist_per_trip'] = city_fare_summary['total_distance'] / city_fare_summary['total_trips']
city_fare_summary = city_fare_summary.merge(dimCity,on='city_id', how='left')

fare_per_kilometer= city_fare_summary['avg_fare_per_trip']/ city_fare_summary['avg_dist_per_trip']
fare_per_kilometer
city_fare_summary['fare_per_kilometer'] = fare_per_kilometer
#city_fare_summary
#--- Rearranging the columns
city_fare_summary = city_fare_summary[['city_id','city_name','total_distance','total_fare','total_trips','avg_fare_per_trip','avg_dist_per_trip','fare_per_kilometer']]
city_fare_summary

city_id      city_name  total_distance  total_fare  total_trips  \
0    AP01  Visakhapatnam          639765     8018282        28366   
1    CH01     Chandigarh          916783    11058401        38981   
2    GJ01          Surat          603122     6431599        54843   
3    GJ02       Vadodara          368867     3797200        32026   
4    KA01         Mysore          267877     4054745        16238   
5    KL01          Kochi         1220167    16997596        50702   
6    MP01         Indore          700629     7635228        42456   
7    RJ01         Jaipur         2308418    37207497        76888   
8    TN01     Coimbatore          316121     3523992        21104   
9    UP01        Lucknow          804571     9463551        64299   

   avg_fare_per_trip  avg_dist_per_trip  fare_per_kilometer  
0         282.672284          22.553938           12.533168  
1         283.686950          23.518714           12.062179  
2         117.272925          10.997247           10.663844  
3         118.566165          11.517736           10.294225  
4         249.707168          16.496921           15.136593  
5         335.245079          24.065461           13.930549  
6         179.838609          16.502473           10.897676  
7         483.918128          30.023125           16.118180  
8         166.982183          14.979198           11.147605  
9         147.180376          12.512963           11.762232

In [16]:
# Sort by Cities with highest fare per kilometer to lowest fare per kilometer.
city_fare_summary = city_fare_summary.sort_values(by='fare_per_kilometer', ascending=False)
city_fare_summary

city_id      city_name  total_distance  total_fare  total_trips  \
7    RJ01         Jaipur         2308418    37207497        76888   
4    KA01         Mysore          267877     4054745        16238   
5    KL01          Kochi         1220167    16997596        50702   
0    AP01  Visakhapatnam          639765     8018282        28366   
1    CH01     Chandigarh          916783    11058401        38981   
9    UP01        Lucknow          804571     9463551        64299   
8    TN01     Coimbatore          316121     3523992        21104   
6    MP01         Indore          700629     7635228        42456   
2    GJ01          Surat          603122     6431599        54843   
3    GJ02       Vadodara          368867     3797200        32026   

   avg_fare_per_trip  avg_dist_per_trip  fare_per_kilometer  
7         483.918128          30.023125           16.118180  
4         249.707168          16.496921           15.136593  
5         335.245079          24.065461           13.930549  
0         282.672284          22.553938           12.533168  
1         283.686950          23.518714           12.062179  
9         147.180376          12.512963           11.762232  
8         166.982183          14.979198           11.147605  
6         179.838609          16.502473           10.897676  
2         117.272925          10.997247           10.663844  
3         118.566165          11.517736           10.294225

Observations:-
- Jaipur and Mysore are the cities with the highest fare per kilometer
- Surat and Vadodra are the citis with the lowest fare per kilometer

3. Average Ratings by City and Fare Type
- Calculate the average passenger and driver ratings for each city, segmented by passenger type (new vs repeate). Identify cities with highest and lowest average ratings.

In [17]:
ratingsSummary = factTrips.groupby(['city_id','passenger_type']).agg(
    avg_passenger_rating = ("passenger_rating",'mean'),
    avg_driver_rating=('driver_rating', 'mean')

).reset_index()

# concatinating the city names column from the dimCity dataset.
ratingsSummary = ratingsSummary.merge(dimCity, on= 'city_id', how= 'left')
ratingsSummary
#---
## Find cities with highest and lowest average passenger ratings
highest_passenger_rating_city = ratingsSummary.loc[ratingsSummary['avg_passenger_rating'].idxmax()]
lowest_passenger_rating_city = ratingsSummary.loc[ratingsSummary['avg_passenger_rating'].idxmin()]

# Find cities with highest and lowest average driver ratings
highest_driver_rating_city = ratingsSummary.loc[ratingsSummary['avg_driver_rating'].idxmax()]
lowest_driver_rating_city = ratingsSummary.loc[ratingsSummary['avg_driver_rating'].idxmin()]
#...
ratingsSummary



city_id passenger_type  avg_passenger_rating  avg_driver_rating  \
0     AP01            new              8.976151           8.979995   
1     AP01       repeated              7.989628           8.992701   
2     CH01            new              8.489158           7.992120   
3     CH01       repeated              7.493798           7.472824   
4     GJ01            new              7.984173           6.994925   
5     GJ01       repeated              5.995511           6.479441   
6     GJ02            new              7.979263           7.004147   
7     GJ02       repeated              5.978629           6.481072   
8     KA01            new              8.982964           8.982878   
9     KA01       repeated              7.978495           8.965767   
10    KL01            new              8.987394           8.985350   
11    KL01       repeated              8.003665           8.989830   
12    MP01            new              8.485837           7.970800   
13    MP01       repeated              7.473961           7.477404   
14    RJ01            new              8.985018           8.988246   
15    RJ01       repeated              7.991042           8.984790   
16    TN01            new              8.485788           7.990604   
17    TN01       repeated              7.475457           7.480778   
18    UP01            new              7.977429           6.990406   
19    UP01       repeated              5.985741           6.491663   

        city_name  
0   Visakhapatnam  
1   Visakhapatnam  
2      Chandigarh  
3      Chandigarh  
4           Surat  
5           Surat  
6        Vadodara  
7        Vadodara  
8          Mysore  
9          Mysore  
10          Kochi  
11          Kochi  
12         Indore  
13         Indore  
14         Jaipur  
15         Jaipur  
16     Coimbatore  
17     Coimbatore  
18        Lucknow  
19        Lucknow

In [18]:
# Display the results
#print("Average Ratings by City and Passenger Type:")
#print(ratingsSummary)

print("\nCity with the Highest Average Passenger Rating:")
print(highest_passenger_rating_city[['city_name', 'passenger_type', 'avg_passenger_rating']])

print("\nCity with the Lowest Average Passenger Rating:")
print(lowest_passenger_rating_city[['city_name', 'passenger_type', 'avg_passenger_rating']])

print("\nCity with the Highest Average Driver Rating:")
print(highest_driver_rating_city[['city_name', 'passenger_type', 'avg_driver_rating']])

print("\nCity with the Lowest Average Driver Rating:")
print(lowest_driver_rating_city[['city_name', 'passenger_type', 'avg_driver_rating']])


City with the Highest Average Passenger Rating:
city_name                  Kochi
passenger_type               new
avg_passenger_rating    8.987394
Name: 10, dtype: object

City with the Lowest Average Passenger Rating:
city_name               Vadodara
passenger_type          repeated
avg_passenger_rating    5.978629
Name: 7, dtype: object

City with the Highest Average Driver Rating:
city_name            Visakhapatnam
passenger_type            repeated
avg_driver_rating         8.992701
Name: 1, dtype: object

City with the Lowest Average Driver Rating:
city_name               Surat
passenger_type       repeated
avg_driver_rating    6.479441
Name: 5, dtype: object


In [19]:
# Group by city_id and calculate average ratings
ratingsSummary = factTrips.groupby('city_id').agg(
    avg_passenger_rating=('passenger_rating', 'mean'),
    avg_driver_rating=('driver_rating', 'mean')
).reset_index()

# Concatenating the city names column from the dimCity dataset
ratingsSummary = ratingsSummary.merge(dimCity, on='city_id', how='left')

# Sorting cities by passenger and driver ratings
top_5_passenger_rating = ratingsSummary.sort_values(by='avg_passenger_rating', ascending=False).head(5)
bottom_5_passenger_rating = ratingsSummary.sort_values(by='avg_passenger_rating', ascending=True).head(5)

top_5_driver_rating = ratingsSummary.sort_values(by='avg_driver_rating', ascending=False).head(5)
bottom_5_driver_rating = ratingsSummary.sort_values(by='avg_driver_rating', ascending=True).head(5)

# Displaying the results
print("Top 5 Cities by Passenger Rating:")
print(top_5_passenger_rating[['city_name', 'avg_passenger_rating']])

print("\nBottom 5 Cities by Passenger Rating:")
print(bottom_5_passenger_rating[['city_name', 'avg_passenger_rating']])

print("\nTop 5 Cities by Driver Rating:")
print(top_5_driver_rating[['city_name', 'avg_driver_rating']])

print("\nBottom 5 Cities by Driver Rating:")
print(bottom_5_driver_rating[['city_name', 'avg_driver_rating']])


Top 5 Cities by Passenger Rating:
       city_name  avg_passenger_rating
4         Mysore              8.701072
7         Jaipur              8.583849
5          Kochi              8.516193
0  Visakhapatnam              8.432948
1     Chandigarh              7.976604

Bottom 5 Cities by Passenger Rating:
    city_name  avg_passenger_rating
2       Surat              6.417081
9     Lucknow              6.489401
3    Vadodara              6.611253
6      Indore              7.828199
8  Coimbatore              7.883055

Top 5 Cities by Driver Rating:
       city_name  avg_driver_rating
5          Kochi           8.987496
0  Visakhapatnam           8.986991
7         Jaipur           8.986851
4         Mysore           8.978076
1     Chandigarh           7.724712

Bottom 5 Cities by Driver Rating:
    city_name  avg_driver_rating
2       Surat           6.588717
9     Lucknow           6.617786
3    Vadodara           6.646475
6      Indore           7.650132
8  Coimbatore           7.6864

Observations :-
- Passenger  Ratings (ratings provided by the passenger to the driver, and the company as a whole)
    - Kochi had the highest avg Passenger Ratings. And it came from new passengers
    - Vadodra had the lowest avg Passenger Ratings. It came from repeat passengers.
- Driver Ratings (ratings given to passengers by the drivers)
    - Vishakapatnam had the highest avg Driver rating, given to repeat passengers
    - Surat had the lowest avg driver rating, again given to repeat passengers

4. Peak and low demand months by city
- For each city, identify the month with the highest total trips(peak demand) and the month with the lowest total trips(low demand)


In [20]:
passengerSummary.loc[passengerSummary.groupby('city_id')['total_passengers'].idxmax()]

month city_id  new_passengers  repeat_passengers  total_passengers
10  2024-02-01    AP01            2380                790              3170
11  2024-02-01    CH01            4104                853              4957
2   2024-01-01    GJ01            2432               1184              3616
13  2024-02-01    GJ02            2146                610              2756
14  2024-02-01    KA01            2107                183              2290
35  2024-04-01    KL01            4939               1576              6515
16  2024-02-01    MP01            2878               1103              3981
17  2024-02-01    RJ01           10789               1661             12450
8   2024-01-01    TN01            1822                392              2214
19  2024-02-01    UP01            3529               1659              5188

In [21]:
# Identify peak demand (highest total passengers) and low demand (lowest total passengers) months for each city
peak_demand = passengerSummary.loc[
    passengerSummary.groupby('city_id')['total_passengers'].idxmax()
]
low_demand = passengerSummary.loc[
    passengerSummary.groupby('city_id')['total_passengers'].idxmin()
]

# Merge with city names for better readability
peak_demand = peak_demand.merge(dimCity, on='city_id', how='left')
low_demand = low_demand.merge(dimCity, on='city_id', how='left')

# Display results
print("Peak Demand (Highest Total Passengers) by City:")
print(peak_demand[['city_name', 'month', 'total_passengers']])

print("\nLow Demand (Lowest Total Passengers) by City:")
print(low_demand[['city_name', 'month', 'total_passengers']])


Peak Demand (Highest Total Passengers) by City:
       city_name       month  total_passengers
0  Visakhapatnam  2024-02-01              3170
1     Chandigarh  2024-02-01              4957
2          Surat  2024-01-01              3616
3       Vadodara  2024-02-01              2756
4         Mysore  2024-02-01              2290
5          Kochi  2024-04-01              6515
6         Indore  2024-02-01              3981
7         Jaipur  2024-02-01             12450
8     Coimbatore  2024-01-01              2214
9        Lucknow  2024-02-01              5188

Low Demand (Lowest Total Passengers) by City:
       city_name       month  total_passengers
0  Visakhapatnam  2024-06-01              2702
1     Chandigarh  2024-04-01              3285
2          Surat  2024-06-01              3030
3       Vadodara  2024-06-01              1807
4         Mysore  2024-04-01              2072
5          Kochi  2024-06-01              4060
6         Indore  2024-06-01              3152
7         Ja

Observations:-
- Peak Demand:
    -  In 7 of the 10 cities, peak demand is in the month of February, 2 of the cities, peak demand is on January, and for the emaining 1 city peak demand is in April
    - So it can be inferred that February is the month that has the most demand.
- Low Demand:
    - In 6 of the 10 cities, peak demand is in the month of June, 2 cities have least demand in May and the remaining 2 in April.

5. Weekend vs Weekday Trip demand by City

In [22]:
# This is how the facTRips data initially looks like. 
factTrips.head()
# I am now going to add the 'day_type' column from the dim_date dataset to teh factTrips dataset.
# This will help me identify the dates as weekend or weekdays.


trip_id       date city_id passenger_type  \
0  TRPLUC240113d55de2fb 2024-01-13    UP01       repeated   
1  TRPVAD240129a3b6dba8 2024-01-29    GJ02       repeated   
2  TRPCOI240107a42430fb 2024-01-07    TN01       repeated   
3  TRPKOC240325d7601389 2024-03-25    KL01       repeated   
4  TRPVIS2406027be97166 2024-06-02    AP01            new   

   distance_travelled(km)  fare_amount  passenger_rating  driver_rating  
0                      11          158                 5              5  
1                       7           74                 5              5  
2                      11          155                 8              8  
3                      36          427                 9             10  
4                      17          265                 8              8

In [23]:
# First, Ensure both datasets have a common column for merging
dimDate['date'] = pd.to_datetime(dimDate['date'])
factTrips['date'] = pd.to_datetime(factTrips['date'])

# Merge the 'day_type' column from dimDate to factTrips based on the 'date' column
factTrips2 = factTrips.merge(dimDate[['date', 'day_type']], on='date', how='left')

# Display the updated factTrips dataset with the 'day_type' column added
factTrips2.head()


trip_id       date city_id passenger_type  \
0  TRPLUC240113d55de2fb 2024-01-13    UP01       repeated   
1  TRPVAD240129a3b6dba8 2024-01-29    GJ02       repeated   
2  TRPCOI240107a42430fb 2024-01-07    TN01       repeated   
3  TRPKOC240325d7601389 2024-03-25    KL01       repeated   
4  TRPVIS2406027be97166 2024-06-02    AP01            new   

   distance_travelled(km)  fare_amount  passenger_rating  driver_rating  \
0                      11          158                 5              5   
1                       7           74                 5              5   
2                      11          155                 8              8   
3                      36          427                 9             10   
4                      17          265                 8              8   

  day_type  
0  Weekend  
1  Weekday  
2  Weekend  
3  Weekday  
4  Weekend

In [24]:
wdayVsWend = factTrips2.groupby(['city_id','day_type'])['trip_id'].count().reset_index()
wdayVsWend_df = pd.DataFrame(wdayVsWend, columns = ['city_id','day_type', 'trip_id'])
wdayVsWend_df= wdayVsWend_df.rename(columns= {'trip_id': 'total_trips'})
wdayVsWend_df
# here we have the data for the total no of trips in the Weekends and weekdays for each of the 10 cities for the entire duration (Jan to June 2024)

city_id day_type  total_trips
0     AP01  Weekday        15100
1     AP01  Weekend        13266
2     CH01  Weekday        19914
3     CH01  Weekend        19067
4     GJ01  Weekday        37793
5     GJ01  Weekend        17050
6     GJ02  Weekday        20310
7     GJ02  Weekend        11716
8     KA01  Weekday         6424
9     KA01  Weekend         9814
10    KL01  Weekday        22915
11    KL01  Weekend        27787
12    MP01  Weekday        21198
13    MP01  Weekend        21258
14    RJ01  Weekday        32491
15    RJ01  Weekend        44397
16    TN01  Weekday        12576
17    TN01  Weekend         8528
18    UP01  Weekday        49617
19    UP01  Weekend        14682

In [25]:
wdayVsWend_df = wdayVsWend_df.merge(dimCity[['city_name', 'city_id']], on='city_id', how='left')
wdayVsWend_df = wdayVsWend_df[['city_id','city_name','day_type','total_trips']]
wdayVsWend_df.head()

city_id      city_name day_type  total_trips
0    AP01  Visakhapatnam  Weekday        15100
1    AP01  Visakhapatnam  Weekend        13266
2    CH01     Chandigarh  Weekday        19914
3    CH01     Chandigarh  Weekend        19067
4    GJ01          Surat  Weekday        37793

Observations:-
- This data can be used to understand demand variations for each city.
- Some cities have higher dmeand on Weekdays and some have higher demand on Weekends.
- Also, observe the difference. How large or small is the difference between demand in weekdays and weekends


---

6. Repeat Passengers Analysis
- Analyze the frequency of trips taken by the repeat passengers in each city (eg, % of repeat passengers taking 2 trips, 3 trips etc,). Identify which cities contribute most to higher trip frequencies amoong repeat passsengers. 

In [26]:
print (repeatTrip.groupby(['city_id','trip_count']).sum().reset_index())

   city_id trip_count  repeat_passenger_count
0     AP01   10-Trips                      47
1     AP01    2-Trips                    2618
2     AP01    3-Trips                    1275
3     AP01    4-Trips                     510
4     AP01    5-Trips                     278
..     ...        ...                     ...
85    UP01    5-Trips                    1768
86    UP01    6-Trips                    1937
87    UP01    7-Trips                    1087
88    UP01    8-Trips                     617
89    UP01    9-Trips                     183

[90 rows x 3 columns]


C:\Users\rohan\AppData\Local\Temp\ipykernel_14952\428698543.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  print (repeatTrip.groupby(['city_id','trip_count']).sum().reset_index())


In [27]:
# Calculate the total number of repeat passengers for each city
city_total_repeat_passengers = repeatTrip.groupby('city_id')['repeat_passenger_count'].sum().reset_index()
city_total_repeat_passengers.rename(columns={'repeat_passenger_count': 'total_repeat_passengers'}, inplace=True)

# Merge the total repeat passengers back with the repeatTrip dataset
repeatTrip_with_totals = repeatTrip.merge(city_total_repeat_passengers, on='city_id', how='left')

# Calculate the percentage of repeat passengers for each trip frequency in each city
repeatTrip_with_totals['percentage'] = (
    repeatTrip_with_totals['repeat_passenger_count'] / repeatTrip_with_totals['total_repeat_passengers'] * 100
)

# Aggregate data to find which cities contribute most to higher trip frequencies
# Filter for higher trip frequencies (e.g., 5 trips or more)
high_frequency = repeatTrip_with_totals[repeatTrip_with_totals['trip_count'].str.extract(r'(\d+)').astype(int) >= 5]
high_frequency_summary = high_frequency.groupby('city_id')['repeat_passenger_count'].sum().reset_index()
high_frequency_summary = high_frequency_summary.merge(dimCity, on='city_id', how='left')

# Sort by the highest contribution to high frequencies
high_frequency_summary = high_frequency_summary.sort_values('repeat_passenger_count', ascending=False)

# Display the results
print("Percentage of Repeat Passengers by Trip Frequency (City-Wise):")
print(repeatTrip_with_totals[['city_id', 'trip_count', 'percentage']])

print("\nCities Contributing Most to Higher Trip Frequencies (5 trips or more):")
print(high_frequency_summary[['city_name', 'repeat_passenger_count']])

Percentage of Repeat Passengers by Trip Frequency (City-Wise):
    city_id trip_count  percentage
0      AP01   10-Trips    0.137040
1      AP01    2-Trips    6.891151
2      AP01    3-Trips    3.093187
3      AP01    4-Trips    1.037588
4      AP01    5-Trips    0.743931
..      ...        ...         ...
535    UP01    5-Trips    2.834219
536    UP01    6-Trips    2.834219
537    UP01    7-Trips    2.563301
538    UP01    8-Trips    0.864854
539    UP01    9-Trips    0.197979

[540 rows x 3 columns]

Cities Contributing Most to Higher Trip Frequencies (5 trips or more):
Empty DataFrame
Columns: [city_name, repeat_passenger_count]
Index: []


In [28]:
# Check the number of unique cities
num_cities = repeatTrip['city_id'].nunique()
print(f"Number of unique cities: {num_cities}")

# Check the number of unique trip frequencies
num_trip_frequencies = repeatTrip['trip_count'].nunique()
print(f"Number of unique trip frequencies: {num_trip_frequencies}")

# Verify total rows in the dataset
print(f"Expected rows: {num_cities * num_trip_frequencies}")


Number of unique cities: 10
Number of unique trip frequencies: 9
Expected rows: 90


7. Monthly target achievement analysis for Key Metrics.
- For each city, evaluate monthly performance against targets for the following using the datasets from targets_db:
    - total trips
    - new passengers
    - average passenger ratings
- Calculate if each metric met, exceeded or misssed the target and calculate the perecentage differenece.

 